presto-main/src/main/java/com/facebook/presto/operator/JoinHashSupplier.java

```diff
@@ -67,7 +84,33 @@ public double getExpectedHashCollisions()
     @Override
     public JoinHash get()
     {
-        Optional<JoinFilterFunction> filterFunction = filterFunctionFactory.map(factory -> factory.create(session, addresses, channels));
```

Comment:
Grrr... :( This makes it more complicated. I can not build `positionLinks` in this `get` method, because it would be too costly to sort position links every time `get` is called. 

But on the other hand, I can not return `positionLinks` created in the constructor with it's `private final filterFunction` here, because then we hit the same problem you are describing.

The only solution that I can see is to build/sort `positionLinks` in the constructor with temporarily created `filterFunction`, but before returning it in in `get`, this filter function must be overwritten by newly created one.